presto-main/src/main/java/com/facebook/presto/execution/StateMachine.java

```diff
@@ -276,34 +278,22 @@ public void addStateChangeListener(StateChangeListener<T> stateChangeListener)
     /**
      * Wait for the state to not be {@code .equals()} to the specified current state.
      */
-    public Duration waitForStateChange(T currentState, Duration maxWait)
+    @VisibleForTesting
+    public void waitForStateChange(T currentState, Duration maxWait)
             throws InterruptedException
     {
         checkState(!Thread.holdsLock(lock), "Can not wait for state change while holding the lock");
         requireNonNull(currentState, "currentState is null");
         requireNonNull(maxWait, "maxWait is null");
 
-        // don't wait if the state has already changed, or we are in a terminal state
-        if (isPossibleStateChange(currentState)) {
-            return maxWait;
+        try {
+            getStateChange(currentState).get(maxWait.toMillis(), TimeUnit.MILLISECONDS);
         }
-
-        // wait for task state to change
-        long remainingNanos = maxWait.roundTo(NANOSECONDS);
-        long start = System.nanoTime();
-        long end = start + remainingNanos;
-
-        synchronized (lock) {
-            while (remainingNanos > 0 && !isPossibleStateChange(currentState)) {
-                // wait for timeout or notification
-                NANOSECONDS.timedWait(lock, remainingNanos);
-                remainingNanos = end - System.nanoTime();
-            }
+        catch (ExecutionException e) {
+            throw new AssertionError(e.getCause());
```

Comment:
The future returned from `StateMachine.getStateChange()` should never throw an exception. Since this method is only for testing, we want to know if there is ever an exception.